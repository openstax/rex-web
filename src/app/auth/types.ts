export interface User {
  firstName: string;
  lastName: string;
  uuid: string;
  isNotGdprLocation: boolean;
}

export interface State {
  user: User | undefined;
  established: boolean;
}
