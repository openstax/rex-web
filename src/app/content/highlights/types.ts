import { Highlight, HighlightColorEnum, HighlightsSummary, NewHighlight } from '@openstax/highlighter/dist/api';
import { ScrollTarget } from '../../navigation/types';
import { FiltersChange } from '../components/popUp/types';
import { LinkedArchiveTreeNode } from '../types';

export interface SummaryHighlights {
  [locationId: string]: {[pageId: string]: HighlightData[]};
}

export type OrderedSummaryHighlights = Array<{
  location: LinkedArchiveTreeNode,
  pages: Array<{
    pageId: string;
    highlights: HighlightData[];
  }>
}>;

export interface SummaryFilters {
  locationIds: string[];
  colors: HighlightColorEnum[];
}

export interface SummaryFiltersUpdate {
  locations?: FiltersChange<LinkedArchiveTreeNode>;
  colors?: FiltersChange<HighlightColorEnum>;
}

export type CountsPerSource = NonNullable<HighlightsSummary['countsPerSource']>;
export type HighlightColorCounts = CountsPerSource[string];

export type SummaryHighlightsPagination = null | {
  sourceIds: string[];
  page: number;
  perPage: number;
};

export interface State {
  currentPage: {
    pageId: string | null,
    highlights: null | HighlightData[];
    hasUnsavedHighlight: boolean;
    focused?: string;
  };
  summary: {
    open: boolean,
    pagination: SummaryHighlightsPagination,
    totalCountsPerPage: CountsPerSource | null;
    filters: SummaryFilters,
    loading: boolean;
    highlights: SummaryHighlights | null;
  };
}

export interface HighlightScrollTarget extends ScrollTarget {
  type: 'highlight';
  id: string;
}

// scopeId should not be nullable in the swagger
export type NewHighlightPayload = NewHighlight & {scopeId: string};
export type HighlightData = Highlight & {scopeId: string};
