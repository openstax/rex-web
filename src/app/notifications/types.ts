import { ActionType } from 'typesafe-actions';
import * as actions from './actions';

export interface Message {
  id: string;
  html: string;
  dismissable: boolean;
  start_at: string | null;
  end_at: string | null;
  url_regex: string | null;
}

export interface AppMessageNotification {
  type: 'Notification/appMessage';
  payload: Message;
}

export type Messages = Message[];

export type ModalNotification = ActionType<Pick<typeof actions,
  'updateAvailable' | 'acceptCookies' | 'retiredBookRedirect'
>>
  | AppMessageNotification;

export interface ToastNotification extends ToastMeta {
  messageKey: string;
  timestamp: number;
  shouldAutoDismiss: boolean;
}

export interface ToastMeta {
  destination: 'studyGuides' | 'myHighlights' | 'page';
  shouldAutoDismiss?: boolean;
  errorId?: string;
}

export type AnyNotification = ModalNotification | ToastNotification;

export interface State {
  modalNotifications: ModalNotification[];
  toastNotifications: ToastNotification[];
}
