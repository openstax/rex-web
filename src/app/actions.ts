import { createStandardAction } from 'typesafe-actions';

export const receivePageFocus = createStandardAction('app/receivePageFocus')<boolean>();
