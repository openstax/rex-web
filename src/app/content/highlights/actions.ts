import { Highlight, NewHighlight, UpdateHighlightRequest } from '@openstax/highlighter/dist/api';
import { createStandardAction } from 'typesafe-actions';
import { LocationFilters } from '../components/popUp/types';
import {
  CountsPerSource,
  HighlightData,
  SummaryFilters,
  SummaryFiltersUpdate,
  SummaryHighlights,
  SummaryHighlightsPagination,
} from './types';

export const focusHighlight = createStandardAction('Content/Highlights/focus')<string>();
export const clearFocusedHighlight = createStandardAction('Content/Highlights/clear')();
export const createHighlight = createStandardAction('Content/Highlights/create')<NewHighlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const requestDeleteHighlight = createStandardAction('Content/Highlights/requestDelete')<Highlight, {
  locationFilterId: string,
  pageId: string,
}>();
export const receiveDeleteHighlight = createStandardAction('Content/Highlights/receiveDelete')<Highlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const updateHighlight = createStandardAction('Content/Highlights/update')<UpdateHighlightRequest, {
  revertingAfterFailure?: boolean,
  preUpdateData: UpdateHighlightRequest,
  locationFilterId: string,
  pageId: string,
}>();
export const receiveHighlights = createStandardAction(
  'Content/Highlights/receive'
)<{highlights: HighlightData[], pageId: string}>();
export const setAnnotationChangesPending = createStandardAction('Content/Highlights/setAnnotationChangesPending')<
  boolean
>();

export const openMyHighlights = createStandardAction('Content/Highlights/Summary/open')<void>();
export const closeMyHighlights = createStandardAction('Content/Highlights/Summary/close')<void>();
export const initializeMyHighlightsSummary = createStandardAction('Content/Highlights/Summary/init')<void>();

export const printSummaryHighlights = createStandardAction('Content/Highlights/Summary/print')();
export const toggleSummaryHighlightsLoading = createStandardAction('Content/Highlights/Summary/loading')<boolean>();

export const loadMoreSummaryHighlights = createStandardAction('Content/Highlights/Summary/loadMore')();
export const setSummaryFilters = createStandardAction('Content/Highlights/Summary/setFilters')<
  Partial<SummaryFilters>
>();
export const updateSummaryFilters = createStandardAction('Content/Highlights/Summary/updateFilters')<
  Partial<SummaryFiltersUpdate>
>();
export const receiveSummaryHighlights = createStandardAction('Content/Highlights/Summary/receiveHighlights')<
  SummaryHighlights,
  {
    pagination: SummaryHighlightsPagination,
    filters?: SummaryFilters,
    isStillLoading?: boolean
  }
>();
export const receiveHighlightsTotalCounts = createStandardAction(
  'Content/receiveHighlightsTotalCounts'
)<CountsPerSource, LocationFilters>();
