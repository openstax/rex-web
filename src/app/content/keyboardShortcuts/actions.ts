import { createStandardAction } from 'typesafe-actions';

export const openKeyboardShortcutsMenu = createStandardAction('Content/KeyboardShortcuts/Menu/open')<void>();
export const closeKeyboardShortcutsMenu = createStandardAction('Content/KeyboardShortcuts/Menu/close')<void>();
