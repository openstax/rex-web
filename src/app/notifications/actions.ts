import { createStandardAction } from 'typesafe-actions';
import { AnyNotification, Message, ToastMeta } from './types';

export const retiredBookRedirect = createStandardAction('Notification/retiredBookRedirect')();
export const updateAvailable = createStandardAction('Notification/updateAvailable')();
export const acceptCookies = createStandardAction('Notification/acceptCookies')();

export const addToast = createStandardAction('Notification/toasts/add')
  .map((messageKey: string, meta: ToastMeta) => ({
    payload: {
      destination: meta.destination,
      errorId: meta.errorId,
      messageKey,
      shouldAutoDismiss: meta.shouldAutoDismiss === undefined || meta.shouldAutoDismiss,
      timestamp: Date.now(),
    },
  }));

export const receiveMessages = createStandardAction('Notification/receiveMessages')<Message[]>();

export const dismissNotification = createStandardAction('Notification/dismiss')<AnyNotification>();
