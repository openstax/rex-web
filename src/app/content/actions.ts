import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as practiceQuestionsActions from './practiceQuestions/actions';
import * as searchActions from './search/actions';
import * as studyGuidesActions from './studyGuides/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;
export const studyGuides = studyGuidesActions;
export const practiceQuestions = practiceQuestionsActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const receiveBuyPrintConfig = createStandardAction('Content/receiveBuyPrintConfig')<State['buyPrint']>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();
export const receivePageNotFoundId = createStandardAction('Content/receivePageNotFoundId')<string>();

export const openNudgeStudyTools = createStandardAction('Content/openNudgeStudyTools')();
export const closeNudgeStudyTools = createStandardAction('Content/closeNudgeStudyTools')();
