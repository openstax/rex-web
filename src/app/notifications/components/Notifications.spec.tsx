import React from 'react';
import renderer from 'react-test-renderer';
import createTestStore from '../../../test/createTestStore';
import TestContainer from '../../../test/TestContainer';
import { Store } from '../../types';
import { acceptCookies, receiveMessages, updateAvailable } from '../actions';
import ConnectedNotifications from './Notifications';

describe('Notifications', () => {
  let store: Store;

  beforeEach(() => {
    store = createTestStore();
  });

  it('matches snapshot', () => {
    store.dispatch(updateAvailable());

    const component = renderer.create(<TestContainer store={store}>
      <ConnectedNotifications />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot', () => {
    store.dispatch(acceptCookies());

    const component = renderer.create(<TestContainer store={store}>
      <ConnectedNotifications />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot', () => {
    store.dispatch(receiveMessages([{
      dismissable: false,
      end_at: null,
      html: 'asdf',
      id: '1',
      start_at: null,
      url_regex: null,
    }]));

    const component = renderer.create(<TestContainer store={store}>
      <ConnectedNotifications />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot for unknown notification', () => {
    store.getState().notifications.modalNotifications.push({type: 'foobar'} as any);

    const component = renderer.create(<TestContainer store={store}>
      <ConnectedNotifications />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
