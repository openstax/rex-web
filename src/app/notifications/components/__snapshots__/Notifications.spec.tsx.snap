// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications matches snapshot 1`] = `
<div
  className="Card__Body-sc-1o7sacr-2 kVpfjE"
>
  <div>
    <div
      className="Card__Group-sc-1o7sacr-0 jzDffS"
    >
      <div
        className="Card__Header-sc-1o7sacr-3 jGwlNH"
      >
        Updates Available
      </div>
      <p
        className="Card__P-sc-1o7sacr-1 bKishz"
      >
        Click below to use the most recent version of this software.
      </p>
    </div>
    <div
      className="Button__ButtonGroup-ayg7nk-1 fJPuTt"
    >
      <button
        className="Button-ayg7nk-0 fFJxez"
        onClick={[Function]}
      >
        Update
      </button>
    </div>
  </div>
</div>
`;

exports[`Notifications matches snapshot 2`] = `
<div
  className="Card__Body-sc-1o7sacr-2 kVpfjE"
>
  <div>
    <div
      className="Card__Group-sc-1o7sacr-0 jzDffS"
    >
      <div
        className="Card__Header-sc-1o7sacr-3 jGwlNH"
      >
        Privacy and cookies
      </div>
      <p
        className="Card__P-sc-1o7sacr-1 bKishz"
        dangerouslySetInnerHTML={
          Object {
            "__html": "This site uses cookies to optimize your experience on our platform. By continuing to use this site you agree to use our cookies. If you'd like more information or would like to opt out please see our <a href=\\"/privacy-policy\\" target=\\"blank\\">Privacy Policy</a>",
          }
        }
      />
      <div
        className="Button__ButtonGroup-ayg7nk-1 fJPuTt"
      >
        <button
          className="Button-ayg7nk-0 fFJxez"
          onClick={[Function]}
        >
          Got it!
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Notifications matches snapshot 3`] = `
<div
  className="Card__Body-sc-1o7sacr-2 kVpfjE"
>
  <div>
    <div
      className="Card__Group-sc-1o7sacr-0 jzDffS"
    >
      <div
        className="Card__Header-sc-1o7sacr-3 jGwlNH"
      >
        Notification
      </div>
      <p
        className="Card__P-sc-1o7sacr-1 bKishz"
        dangerouslySetInnerHTML={
          Object {
            "__html": "asdf",
          }
        }
      />
    </div>
    <div
      className="Button__ButtonGroup-ayg7nk-1 fJPuTt"
    >
      <button
        className="Button-ayg7nk-0 fFJxez"
        onClick={[Function]}
      >
        OK
      </button>
    </div>
  </div>
</div>
`;

exports[`Notifications matches snapshot for unknown notification 1`] = `null`;
